### ISSUE TITLE
Unexpected result of key joining when key is not ordered.

### ISSUE THREAD

**Issue by realzhang** [Rdatatable/data.table#5657]
library(data.table)
library(tidyverse)

dt1 <- data.table(chr = "chr1",
                  pos = 1:8,
                  val = 1:8 *100) %>% setkeyv(c("chr", "pos"))

dt2 <- data.table(chr = "chr1",
                  pos = c(3,7,9,11),
                  val = c(3,7,9,11) *200) %>% setkeyv(c("chr", "pos"))

allKey <- dplyr::union(dt2[, .(chr,pos)], dt1[, .(chr,pos)]) %>% setkeyv(c("chr", "pos")) # why setkeyv not sort allKey when dt2 is the first parameter.
#allKey <- dplyr::union(dt1[, .(chr,pos)], dt2[, .(chr,pos)]) %>% setkeyv(c("chr", "pos")) # this works as expected.
dt1[allKey] # vals are got.
dt2[allKey] # get nothing overlapping, why?

**Comment by avimallu** [comment]
This is because `dplyr::union` isn't aware of `data.table`, **and** it explicitly preserves attributes. The key of a data.table object is stored as an attribute. In `dplyr::union`:

```r
> dplyr::union(dt2[, .(chr,pos)], dt1[, .(chr,pos)]) |> str()
Classes ‘data.table’ and 'data.frame':	10 obs. of  2 variables:
 $ chr: chr  "chr1" "chr1" "chr1" "chr1" ...
 $ pos: num  3 7 9 11 1 2 4 5 6 8
 - attr(*, ".internal.selfref")=<externalptr> 
 - attr(*, "sorted")= chr [1:2] "chr" "pos"
```

Your output of `allKey` will be marked as "sorted" because `dplyr::union` retains (an incorrect) attribute. `setkeyv` thinks that`allKey` is already sorted, so doesn't bother to sort it again. You'll notice the difference when you use `rbindlist` (which knows to reset the `sorted` attribute):

```r
> rbindlist(list(dt2[, .(chr, pos)], dt1[, .(chr, pos)])) |> str()
Classes ‘data.table’ and 'data.frame':	12 obs. of  2 variables:
 $ chr: chr  "chr1" "chr1" "chr1" "chr1" ...
 $ pos: num  3 7 9 11 1 2 3 4 5 6 ...
 - attr(*, ".internal.selfref")=<externalptr> 
```

Notice how the ` - attr(*, "sorted")= chr [1:2] "chr" "pos"` isn't there in the output. `setkeyv` will therefore work with it.

## TL;DR
`setkeyv` doesn't do any sorting because `dplyr::union` retains the `sorted` attribute in either case. It just happens that the UNION operation on `dt1` followed by `dt2` is already sorted by chance, so it **appears** that `setkeyv` is working, but actually doesn't work.

## Solution
Use `rbindlist` followed by `unique`, or, if you absolutely want to use `dplyr::union`, do:

```r
library(data.table)
library(dplyr)

dt1 <- data.table(chr = "chr1",
                  pos = 1:8,
                  val = 1:8 * 100) %>% setkeyv(c("chr", "pos"))

dt2 <- data.table(chr = "chr1",
                  pos = c(3, 7, 9, 11),
                  val = c(3, 7, 9, 11) * 200) %>% setkeyv(c("chr", "pos"))

setkey(dt1, NULL)
setkey(dt2, NULL)
dplyr::union(dt2[, .(chr,pos)], dt1[, .(chr,pos)]) |> setkeyv(c("chr", "pos")) |> print()
```

This removes the `sorted` attribute in the `data.table` objects.

**Comment by jangorecki** [comment]
Don't we have funion function?

**Comment by realzhang** [comment]
Thanks a lot for the detailed reply! The above two comments are all great!
After I opened this issue, I used` merge(dt1, dt2, all.x=T, all.y=T)` as a work around.
Then I searched alternatives for `dplyr::union`, and yes, I found `funion`.

```
allKey2 <- funion(dt2[, .(chr,pos)], dt1[, .(chr,pos)])
allKey3 <- funion(dt2[, .(chr,pos)], dt1[, .(chr,pos)]) %>% setkeyv(c("chr","pos"))
dt2(allKey2)
dt2(allKey3)
```

All worked like a charm! Cheers~
Thank you all and I'll close it with comment.

**Comment by avimallu** [comment]
> Don't we have funion function?

Oh dear. Have been out of touch with data.table a while. Completely forgot about those fast set operation functions!
